lib/oauth2-client.js

```diff
@@ -22,68 +23,38 @@ bedrock.events.on('bedrock.init', async () => {
   });
 });
 
-export class OAuth2Client {
-  /**
-   * Constructs an OAuth2Client instance to return an access token. It first
-   * checks the database to see if an access token already exists for the client
-   * id; if not, it creates an access token then stores it in the database.
-   *
-   * @param {object} options - Options to use.
-   * @param {string} options.config - The "oauth2-client-grant" properties.
-   * @param {object} options.httpsAgent - A NodeJS HTTPS Agent (`https.Agent`).
-   * @param {string} options.issuer - The issuer.
-   * @param {string} options.logger - The logger.
-   *
-   * @returns {OAuth2Client} The `OAuth2Client` instance.
-   */
-  constructor({config, httpsAgent, issuer, logger} = {}) {
-    const cfg = config.find(a => a.issuer === issuer);
-    this.client_id = cfg.client_id;
-    this.client_secret = cfg.client_secret;
-    this.token_endpoint = cfg.token_endpoint;
-    this.grant_type = cfg.grant_type;
-    this.scope = cfg.scope;
-    this.maxRetries = cfg.maxRetries;
-    this.httpsAgent = httpsAgent;
-    this.logger = logger;
-  }
-
-  /**
+/**
    * Gets a new access token by first checking the database to see if an
    * access token already exists for the client id; if not it creates an
    * access token then stores it in the database.
    *
    * @param {object} options - Options to use.
-   * @param {string} options.retryAndExitOnFailure - If the request should be
-   *  retried indefinitely, and exit upon an unrecoverable error.
+   * @param {string} options.client_id - The ID of the client.
+   * @param {string} options.client_secret - The client secret.
+   * @param {string} options.token_endpoint - The URL to call.
+   * @param {string} options.grant_type - The grant type.
+   * @param {string} options.scope - The scope.
+   * @param {number} options.maxRetries - The maximum number of times to retry
+   *  the request.
+   *
    * @returns {Promise<string>} The access token.
    */
-  async getAccessToken({retryAndExitOnFailure = false}) {
-    const {
-      client_id, client_secret, token_endpoint, grant_type, scope, httpsAgent,
-      maxRetries
-    } = this;
-    if(retryAndExitOnFailure) {
-      return getAccessTokenRetryAndExitOnFailure({
-        client_id, client_secret, token_endpoint, grant_type, scope, httpsAgent
-      });
-    }
-    const fn = () => _getUncachedAccessToken({
-      client_id, client_secret, token_endpoint, grant_type, scope, httpsAgent,
-      maxRetries
-    });
-    return ACCESS_TOKEN_CACHE.memoize({key: client_id, fn});
-  }
+export async function getAccessToken({
+  client_id, client_secret, token_endpoint, grant_type, scope, maxRetries
+}) {
+  const fn = () => _getUncachedAccessToken({
+    client_id, client_secret, token_endpoint, grant_type, scope, maxRetries
+  });
+  return ACCESS_TOKEN_CACHE.memoize({key: client_id, fn});
```

Comment:
@dmitrizagidulin Thoughts?

Reply:
@mandyvenables the key should be whatever uniquely identifies a particular call to get the access token. If the caller ever receives an access token that doesn't match the parameters they passed in an unexpected/acceptable way, then there's a problem.

I think, at a minimum, the `token_endpoint` needs to be included because we mentioned that a client ID is only unique with respect to a particular endpoint. @dmitrizagidulin -- what other characteristics are there here for uniqueness and expectation?

Reply:
So, this is tricky. The `client_id` typically has a 1-1 mapping to the `token_endpoint` (meaning, if you need to go to a different `token_endpoint`, it generally means that it's a different zone of trust, and you'll be issued a different `client_id`). 

(This is, at least, for the typical OAuth2 setup, where mobile clients & third-party apps register with the authorization server, and receive their individual `client_id` and `client_secret` strings. It doesn't really hold for globally unique client_ids (like, WebIDs or DIDs) -- it's just not a use case the community has dealt with much.)

It's not inconceivable -- the protocol does not forbid reusing the client_id on multiple token endpoints.
So, I could see it either way (either just memoizing solely by `client_id`, or the `${client_id}${token_endpoint}` tuple.

Reply:
Yeah, the more I think about it -- let's just memoize by client_id + endpoint.

Reply:
If it's possible that `client_id` may be the same across `token_endpoint`s then it's imperative that it be included in the key, otherwise it's a security risk. Additionally, the key must have a delimiter that isn't present in either of those values (presuming that they are the only values that need to go into the key, I'm still not sure). So, whatever values need to go into the key should probably just be bundled up using `JSON.stringify([a, b, c...])` unless we've got a better idea for delimiting.

Back to uniqueness -- do any of the other parameters need to be considered as part of the key?

Reply:
@dlongley I think the combination of `client_id` and `token_endpoint` should be sufficient, I can't think of any use cases where there would be identical `client_id`s  with different secrets on the same endpoint.

Reply:
Ok, so, let's examine all those parameters.
I think we agree at this point that `client_id` + `token_endpoint` is our initial unique caching key. Let's look at the others.

`client_secret` -- not useful. It's not allowed to have multiple registrations for the same `client_id` but with different `client_secret`. In the _rare_ occurrence that a client re-registers with the same client id (or updates their secret), then the AS (authorization server) will either accept the old token (in which case, the cache will work without `client_secret`), or it won't (in which case, the cache doesn't work, and the token gets re-requested -- a price of another http roundtrip, but again, this is a rare occurrence that we don't need to optimize for, cache-wise).

`grant_type` -- not useful. Grant type has to do with how the client authenticated / requested the token in the first place. But by the time the logic gets down to issuing the token, the grant type is forgotten / does not matter. Put a different way, a client could reasonably get the same token regardless of which grant_type it used to request it.

`scope` -- This one's the closest candidate, as far as our composite caching key. Again, though, also not that useful. (Our client management libraries (or anybody's, really), don't juggle different tokens (from the same issuer) based on scope.

Reply:
> (Our client management libraries (or anybody's, really), don't juggle different tokens (from the same issuer) based on scope.

Heh, the ambient authority and blast surface damage problems with oauth 2 in a nutshell. :P